File: options_trader_node/src/utils/greeks.js
import { toFloat } from "../scanner/filters.js";

export function averageTrueRange(candles, period = 14) {
  if (!Array.isArray(candles) || candles.length < 2) return 0;

  const highs = candles.map((c) => toFloat(c.highPrice ?? c.high)).filter((v) => v > 0);
  const lows = candles.map((c) => toFloat(c.lowPrice ?? c.low)).filter((v) => v >= 0);
  const closes = candles.map((c) => toFloat(c.closePrice ?? c.close)).filter((v) => v > 0);

  if (highs.length < 2 || lows.length < 2 || closes.length < 2) return 0;

  const n = Math.min(highs.length, lows.length, closes.length);
  const trueRanges = [];
  for (let i = 1; i < n; i += 1) {
    const tr = Math.max(
      highs[i] - lows[i],
      Math.abs(highs[i] - closes[i - 1]),
      Math.abs(lows[i] - closes[i - 1]),
    );
    trueRanges.push(Math.max(tr, 0));
  }

  if (trueRanges.length === 0) return 0;
  const window = trueRanges.slice(-period);
  return window.reduce((a, b) => a + b, 0) / window.length;
}

File: options_trader/utils/greeks.py
from __future__ import annotations

import math
from typing import Any


def average_true_range(candles: list[dict[str, Any]], period: int = 14) -> float:
    if len(candles) < 2:
        return 0.0

    highs: list[float] = []
    lows: list[float] = []
    closes: list[float] = []
    for c in candles:
        try:
            highs.append(float(c.get("high_price") or c.get("high")))
            lows.append(float(c.get("low_price") or c.get("low")))
            closes.append(float(c.get("close_price") or c.get("close")))
        except (TypeError, ValueError):
            continue

    if len(closes) < 2:
        return 0.0

    true_ranges: list[float] = []
    for i in range(1, len(closes)):
        high = highs[i]
        low = lows[i]
        prev_close = closes[i - 1]
        tr = max(high - low, abs(high - prev_close), abs(low - prev_close))
        true_ranges.append(max(tr, 0.0))

    if not true_ranges:
        return 0.0
    window = true_ranges[-period:] if len(true_ranges) > period else true_ranges
    return sum(window) / len(window)


def _norm_cdf(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def black_scholes_delta(
    spot: float,
    strike: float,
    time_years: float,
    rate: float,
    volatility: float,
    option_type: str,
) -> float:
    if min(spot, strike, time_years, volatility) <= 0:
        return 0.0
    d1 = (math.log(spot / strike) + (rate + 0.5 * volatility**2) * time_years) / (volatility * math.sqrt(time_years))
    if option_type.lower() == "call":
        return _norm_cdf(d1)
    return _norm_cdf(d1) - 1.0
